### PR TITLE
Assignment 1 & 2 - Scenario : I should be able to fetch all posts, and all comments for user with id '1' &&  Scenario: User should be able to fetch all the albums for user id 2 && Scenario 3: User should be able to create a post and then edit the created post

### PR DESCRIPTION
**Pull request by AmanDaharwal**

### CODE REVIEW

File: src/test/java/com/znsio/api/posts.feature [hunk 1]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/api/posts.feature]: Feature name doesn't match what tests are inside it. It has comment Api tests also.

File: src/test/java/com/znsio/api/posts.feature [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  @posts @prod
+  Feature: Api tests for https://jsonplaceholder.typicode.com/
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/api/posts.feature]: Prefer not to mention URL in  here. Please use appropriate feature description.

File: src/test/java/com/znsio/templates/postsTemplates.feature [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      Given path "/posts/" + id
+      And print "Get the posts for a user from :"+env.postsUrl
+      When method get
+      Then status 200
+      And def postsForUser = response
+      And print "Response for get posts for a user " + response
+
+    @t_getCommentsForUser
+    Scenario: Get all the comments for user 1
+      Given path "/posts/"+ postId +"/comments"
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/templates/postsTemplates.feature]: Why we are fetching comments based on postId? Assignment is to fetch comments for user 1. Let's discuss if you have any doubts .

File: src/test/java/com/znsio/templates/postsTemplates.feature [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      And print "Get all the comments for user from "+env.postUrl
+      When method get
+      Then status 200
+      And def postsForUser = response
+      And print "Response for get comments for user " + response
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/templates/postsTemplates.feature]: Use karate.log() in place of print statement.

File: src/test/java/com/znsio/api/posts.feature [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+    Scenario: User should be able to fetch all the posts for a user with id 1
+      * def id = 1
+      Given def postForUser = karate.call('classpath:com/znsio/templates/postsTemplates.feature@t_getPostsWithId', { 'id' : id }).postsForUser
+      Then assert postForUser.id == id
+      And match postForUser.id == '#number'
+
+    Scenario: User should be able to fetch all the comments for user 1
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/api/posts.feature]: Please check templates comment for this scenario

File: src/test/java/com/znsio/templates/createPostPayload.json [hunk 6]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/templates/createPostPayload.json]: How will you manage multiple data json here?

File: src/test/java/com/znsio/templates/updatePostPayload.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/templates/updatePostPayload.json]: Can be defined in single JSONs file where all template data is available.

File: src/test/java/com/znsio/templates/updatePostPayload.json [hunk 8]
```diff
@@ -0,0 +1,6 @@
...
+  "title": "djhakds",
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/templates/updatePostPayload.json]: Use randomiser feature to generate random string! would be more efficient.

File: src/test/java/test_data.json [hunk 9]
```diff
@@ -20,7 +20,8 @@
       "baseUrl": "https://github.com",
       "confEngineUrl": "https://confengine.com/api/v3",
       "dummyRestAPIUrl": "https://jsonplaceholder.typicode.com",
-      "restCountryUrl": "https://restcountries.com/v3.1"
+      "restCountryUrl": "https://restcountries.com/v3.1",
+      "postsUrl": "https://jsonplaceholder.typicode.com"
```
Comment (mshahrukh2) [inline src/test/java/test_data.json]: Wrong url naming convention.

File: src/test/java/com/znsio/api/posts.feature [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+      * def postId = 8
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/api/posts.feature]: Please go through scenario again.

**Review by mshahrukh2 — CHANGES_REQUESTED** [review]
Please incorporate review comments!

File: src/test/java/test_data.json (hunk 9 above)
Comment (AmanDaharwal) [reply src/test/java/test_data.json]: Done, updated

File: src/test/java/com/znsio/api/posts.feature (hunk 1 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/api/posts.feature]: updated name

File: src/test/java/com/znsio/api/posts.feature (hunk 2 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/api/posts.feature]: removed url

File: src/test/java/com/znsio/templates/postsTemplates.feature (hunk 4 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/templates/postsTemplates.feature]: Done

File: src/test/java/com/znsio/api/jsonPlaceHolderSchema.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/api/jsonPlaceHolderSchema.json]: This File is located in wrong directory.

File: src/test/java/com/znsio/api/jsonPlaceHolder.feature [hunk 12]
```diff
@@ -0,0 +1,39 @@
+
+  @jsonplaceholder @prod
+  Feature: Api tests for Json Place Holder
+
+    Background:
+      * def testSchema = read('classpath:com/znsio/api/jsonPlaceHolderSchema.json')
+
+    @posts
+    Scenario Outline: User should be able to fetch all the posts for a user with id <id>
+      * def id = 10
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/api/jsonPlaceHolder.feature]: * def id = 10 -Is this required?

File: src/test/java/com/znsio/templates/jsonPlaceHolderTemplate.feature [hunk 13]
```diff
@@ -0,0 +1,56 @@
+
+@posts
+Feature: Api tests for Json Place Holder
+
+  Background:
+    Given url env.jsonPlaceHolderUrl
+    And print "Post Url : "+env.jsonPlaceHolderUrl
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/templates/jsonPlaceHolderTemplate.feature]: Why do we need to print URL?

File: src/test/java/com/znsio/workflow/jsonPlaceHolderTemplate.feature [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+    Background:
+      * def userId = generateRandomNumber(3)
+      * def title = generateAlphaNumericRandomString(5)
+      * def body = generateAlphaNumericRandomString(10)
+      * def jsonPayload = { "userId": "#(userId)", "title": "#(title)", "body": "#(body)" }
+      * karate.log("jsonPayload : "+jsonPayload.userId+" "+jsonPayload.title+" "+jsonPayload.body)
+
+    Scenario: User should be able to create a post and then edit the created post
+      Given def postedUserResponse = karate.call('classpath:com/znsio/templates/postsTemplates.feature@t_postPostsForUser', {'jsonPayload' : jsonPayload , 'expectedStatus': 201}).postForUser
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/workflow/jsonPlaceHolderTemplate.feature]: This is wrong !. First verify the Response of fist API call then if its working ok then call another API.

File: src/test/java/com/znsio/workflow/jsonPlaceHolderTemplate.feature [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+      And def fetchedUserResponse = karate.call('classpath:com/znsio/templates/postsTemplates.feature@t_getPostsWithId', { 'id' : postedUserResponse.id, 'expectedStatus' : 200 }).postsForUser
+      Then assert postedUserResponse.id == fetchedUserResponse.id
+      And assert postedUserResponse.title == fetchedUserResponse.title
+      And assert postedUserResponse.body == fetchedUserResponse.body
+      * def newTitle = generateAlphaNumericRandomString(5)
+      * def newJsonPayload = { "userId": "#(userId)", "title": "#(newTitle)", "body": "#(body)" }
+      When def updatedUserResponse = karate.call('classpath:com/znsio/templates/postsTemplates.feature@t_patchPostsForUser', { 'id' : postedUserResponse.id, 'jsonPayload' : newJsonPayload, 'expectedStatus' : 200 }).postsForUser
+      And def postedUserResponse2 = karate.call('classpath:com/znsio/templates/postsTemplates.feature@t_postPostsForUser', 'expectedStatus' : 200 ).postForUser
```
Comment (mshahrukh2) [inline src/test/java/com/znsio/workflow/jsonPlaceHolderTemplate.feature]: Same as above

**Pull request closed without merging**